pkg/modules/modules.go

```diff
@@ -149,3 +164,31 @@ func (m *service) RegisterInvisibleModule(name string, initFn func() (services.S
 func (m *service) IsModuleEnabled(name string) bool {
 	return stringsContain(m.targets, name)
 }
+
+// processFeatureFlags adds or removes targets based on feature flags.
+func (m *service) processFeatureFlags() error {
+	// add GrafanaAPIServer to targets if feature is enabled
+	if m.features.IsEnabled(featuremgmt.FlagGrafanaAPIServer) {
+		m.targets = append(m.targets, GrafanaAPIServer)
+	}
+
+	if !m.features.IsEnabled(featuremgmt.FlagGrafanaAPIServer) {
+		// error if GrafanaAPIServer is in targets
+		for _, t := range m.targets {
+			if t == GrafanaAPIServer {
+				return fmt.Errorf("feature flag %s is disabled, but target %s is still enabled", featuremgmt.FlagGrafanaAPIServer, GrafanaAPIServer)
+			}
+		}
+
+		// error if GrafanaAPIServer is a dependency of a target
+		for parent, targets := range dependencyMap {
+			for _, t := range targets {
+				if t == GrafanaAPIServer && m.IsModuleEnabled(parent) {
+					return fmt.Errorf("feature flag %s is disabled, but target %s is enabled with dependency on %s", featuremgmt.FlagGrafanaAPIServer, parent, GrafanaAPIServer)
+				}
+			}
+		}
+	}
+
+	return nil
```

Comment:
@ryantxu does this logic seem ok for handling the feature flag?

Reply:
For this PR -- i think this is sufficient.  But I think we want to make a more clear relationship between targests and featureflags.  I am exploring that in a branch, but it is not complicated for this PR :)